src/index.py

```diff
@@ -79,3 +96,260 @@ def contents_from_page(page: str) -> str:
+        docs_path: The base directory of all items.
+
+    Raises:
+        InputError:
+            - An item has more whitespace than a previous item and it is not following a directory.
+            - A nested item is not immediately within the path of its parent.
+            - An item isn't a file nor directory.
+    """
+    # Check that the whitespace count matches the expectation
+    if item.whitespace_count > whitespace_expectation:
+        raise InputError(
+            "An item has more whitespace and is not following a reference to a directory. "
+            f"{item=!r}, expected whitespace count: {whitespace_expectation!r}"
+        )
+
+    # Check that the next item is within the directory
+    item_path = Path(item.reference_value)
+    try:
+        item_to_aggregate_path = item_path.relative_to(aggregate_dir)
+    except ValueError as exc:
+        raise InputError(
+            "A nested item is a reference to a path that is not within the directory of its "
+            f"parent. {item=!r}, expected parent path: {aggregate_dir!r}"
+        ) from exc
+
+    # Check that the item is directly within the current directory
+    if len(item_to_aggregate_path.parents) != 1:
+        raise InputError(
+            "A nested item is a reference to a path that is not immediately within the "
+            f"directory of its parent. {item=!r}, expected parent path: {aggregate_dir!r}"
+        )
+
+    # Check that if the item is a file, it has the correct extension
+    if (item_path := docs_path / Path(item.reference_value)).is_file():
+        if item_path.suffix.lower() != DOC_FILE_EXTENSION:
+            raise InputError(
+                "An item in the contents list is not of the expected file type. "
+                f"{item=!r}, expected extension: {DOC_FILE_EXTENSION}"
+            )
+
+
+def _calculate_contents_hierarchy(
+    parsed_items: "peekable[_ParsedListItem]",
+    docs_path: Path,
+    aggregate_dir: Path = Path(),
+    hierarchy: int = 0,
+    whitespace_expectation: int = 0,
+) -> typing.Iterator[IndexContentsListItem]:
+    """Calculate the hierarchy of the contents list items.
+
+    Args:
+        parsed_items: The parsed items from the contents list in the index file.
+        docs_path: The base directory of all items.
+        aggregate_dir: The relative directory that all items must be within.
+        hierarchy: The hierarchy of the current directory.
+        whitespace_expectation: The expected number of whitespace characters for items.
+
+    Yields:
+        The contents list items with the hierarchy.
+
+    Raises:
+        InputError:
+            - An item has more whitespace than a previous item and it is not following a directory.
+            - A nested item is not immediately within the path of its parent.
+            - An item isn't a file nor directory.
+    """
+    while next_item := parsed_items.peek(default=None):
```

Comment:
because the item should not be consumed from the iterator if the `next_item.whitespace_count < whitespace_expectation` condition is met since it should be processed by the previous level of recursion

Reply:
What about this approach ?
https://github.com/gregory-schiano/upload-charm-docs/commit/86e31c426d4d0debe763a49d0703ce8eec914520

Reply:
And using this approach you can easily add whitespace coherence check:
https://github.com/gregory-schiano/upload-charm-docs/commit/00542a85bb643bf5d2298b15292380f1350889ae

Even number/letters coherence will be easier

Reply:
I think the idea behind these is fine, made a few adjustments and pushed this